### ISSUE TITLE
How to reproduce the rec performance on INSPIRED dataset?

### ISSUE THREAD

**Issue by dayuyang1999** [wxl1999/UniCRS#7]
Dear Author,

I am trying to reproduce the rec performance on INSPIRED dataset.

![image](https://github.com/wxl1999/UniCRS/assets/60718476/4fb55643-4784-4995-8271-fa1479cd766b)

I use the hyperparameters you recommend and the "best" model as prompt-encoder. Unfortunately, I was not able to reproduce the performance on the paper.

---- Here I attached the loss and recall@1 on testset for prompt pre-training, conversational training, and recommendation training steps:
![image](https://github.com/wxl1999/UniCRS/assets/60718476/20d5f89c-9ba0-4f42-b965-9db7b4e539b4)
![image](https://github.com/wxl1999/UniCRS/assets/60718476/dcf74f86-fa4f-4d6f-99e7-999e883b8cc7)

 prompt pre-training

![image](https://github.com/wxl1999/UniCRS/assets/60718476/61f17289-3696-4b95-a3b6-b1a0326fa676)
conversational training

![image](https://github.com/wxl1999/UniCRS/assets/60718476/011f53fa-57ff-4a10-8ac2-76bdff5c9f57)
![image](https://github.com/wxl1999/UniCRS/assets/60718476/9c7425b7-9a9f-43a4-bc1b-b944b09af74a)
recommendation training (as you can see, the best recall@1 I got is around 0.04, far from 0.09)

---- and here are the configuration I use for prompt pre-training, conversational training, and recommendation training steps:

```
python3 train_pre.py \
    --dataset inspired \
    --tokenizer microsoft/DialoGPT-small \
    --model microsoft/DialoGPT-small \
    --text_tokenizer roberta-base \
    --text_encoder roberta-base \
    --num_train_epochs 5 \
    --gradient_accumulation_steps 1 \
    --per_device_train_batch_size 64 \
    --per_device_eval_batch_size 128 \
    --num_warmup_steps 168 \
    --max_length 200 \
    --prompt_max_length 200 \
    --entity_max_length 32 \
    --learning_rate 6e-4 \
    --output_dir UniCRS/src/result_promptpretraining_inspired \
    --use_wandb \
    --project crs-prompt-pre-inspired \
    --name exp1 \
    --gpu 0 
```
prompt pre-training

```
python3 train_conv.py \
    --dataset inspired \
    --tokenizer microsoft/DialoGPT-small \
    --model microsoft/DialoGPT-small \
    --text_tokenizer roberta-base \
    --text_encoder roberta-base \
    --n_prefix_conv 20 \
    --prompt_encoder UniCRS/src/result_promptpretraining_inspired/best/ \
    --num_train_epochs 10 \
    --gradient_accumulation_steps 1 \
    --ignore_pad_token_for_loss \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 16 \
    --num_warmup_steps 976 \
    --context_max_length 200 \
    --resp_max_length 183 \
    --prompt_max_length 200 \
    --entity_max_length 32 \
    --learning_rate 1e-4 \
    --output_dir UniCRS/src/result_convprompt_inspired \
    --use_wandb \
    --project crs-prompt-conv-inspired \
    --name exp1 \
    --gpu 0

```
conv training

```
python3 infer_conv.py \
    --dataset inspired \
    --split test \
    --tokenizer microsoft/DialoGPT-small \
    --model microsoft/DialoGPT-small \
    --text_tokenizer roberta-base \
    --text_encoder roberta-base \
    --n_prefix_conv 20 \
    --prompt_encoder UniCRS/src/result_convprompt_inspired/best \
    --per_device_eval_batch_size 64 \
    --context_max_length 200 \
    --resp_max_length 183 \
    --prompt_max_length 200 \
    --entity_max_length 32 \
    --gpu 1
```
conv infer

```
python3 train_rec.py \
    --dataset inspired_gen \
    --tokenizer microsoft/DialoGPT-small \
    --model microsoft/DialoGPT-small \
    --text_tokenizer roberta-base \
    --text_encoder roberta-base \
    --n_prefix_rec 10 \
    --prompt_encoder UniCRS/src/result_promptpretraining_inspired/best \
    --num_train_epochs 5 \
    --per_device_train_batch_size 64 \
    --per_device_eval_batch_size 64 \
    --gradient_accumulation_steps 1 \
    --num_warmup_steps 33 \
    --context_max_length 200 \
    --prompt_max_length 200 \
    --entity_max_length 32 \
    --learning_rate 1e-4 \
    --output_dir UniCRS/src/result_rec_inspired \
    --use_wandb \
    --project crs-prompt-rec-inspired \
    --name exp1 \
    --gpu 0
```

rec training

Thank you!

**Comment by wxl1999** [comment]
Sorry for the late reply! The problem comes from the pre-training stage. You should observe a continuous increase in performance and drop in loss since the answer is actually provided in the response.

- loss
![image](https://github.com/wxl1999/UniCRS/assets/47387759/b3571bf0-ab13-48f5-8277-3d1eb13dff9c)

- Recall
![image](https://github.com/wxl1999/UniCRS/assets/47387759/ed99f18f-73a7-4e6f-8f6e-bb194dde806c)

**Comment by invoker-LL** [comment]
> Sorry for the late reply! The problem comes from the pre-training stage. You should observe a continuous increase in performance and drop in loss since the answer is actually provided in the response.
> 
> * loss
>   ![image](https://private-user-images.githubusercontent.com/47387759/285855038-b3571bf0-ab13-48f5-8277-3d1eb13dff9c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE4ODEyMTksIm5iZiI6MTcwMTg4MDkxOSwicGF0aCI6Ii80NzM4Nzc1OS8yODU4NTUwMzgtYjM1NzFiZjAtYWIxMy00OGY1LTgyNzctM2QxZWIxM2RmZjljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA2VDE2NDE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4NDg1NzAzNzU3NWE1Yzg3MDNlODVkZDQ2Zjk1ZDUxNzk0NWYzNTYyNGVhOTg0NjhhMTU0ZGNlZDAwYzBjN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.b3qVgwolfIp9ridAjkVLcEEObAoPEFtsNLgMCNYCVQ4)
> * Recall
>   ![image](https://private-user-images.githubusercontent.com/47387759/285854997-ed99f18f-73a7-4e6f-8f6e-bb194dde806c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE4ODEyMTksIm5iZiI6MTcwMTg4MDkxOSwicGF0aCI6Ii80NzM4Nzc1OS8yODU4NTQ5OTctZWQ5OWYxOGYtNzNhNy00ZTZmLThmNmUtYmIxOTRkZGU4MDZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA2VDE2NDE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyYjUxYzY0YzBmZjc3YmEzMTZhMjhmZmQ3MjA0NmRhMDQ4ZjhlY2FiNDZjMmE3NzJhMGI0NDkyMTRiMDc2ZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RxGwCq0pWyrMqWajj8-72u8DRYtpw8ODvhcXwVuOM_A)

Though my pretraining results is as good as above, the recommendation result is still lower too much compared to paper results.

pretraining：
![image](https://github.com/wxl1999/UniCRS/assets/35137989/c91e54cc-75fe-44ea-9b93-15fbc680a415)

recommendation: 
![image](https://github.com/wxl1999/UniCRS/assets/35137989/088edd53-c502-4d14-90ab-a34ecab8b131)

But I can easily reproduce the UniCRS results in CFCRS code, what's their difference?

**Comment by wxl1999** [comment]
@invoker-LL Hi, your problem is interesting. The metric `test/loss` is abnormal, since it keeps increasing. You can compare the code with UniCRS for more detail. According to my impression, the main difference is that the length of soft token prompt is set to 0 in CFCRS
